### ISSUE TITLE
Access the wrong 8-bit register when using MMIO

### ISSUE THREAD

**Issue by SimK98l** [intel/haxm#189]
**Describe the Bug**

Summary: 
When accessing the 8-bit register with instructions related to MMIO, HAXM access the wrong register.
In the current implementation of HAXM, register numbers are allocated in the following order.
AL, CL, DL, BL, SP, BP, SI, DI
However, when using the 8-bit register, the following order is correct[1].
AL, CL, DL, BL, AH, CH, DH, BH

[1] IA-32 Intel(R) Architecture Software Developer’s Manual Volume 2: Instruction Set Reference
B.1.1. Reg Field (reg)

**Host Environment**

* HAXM version: 7.4.1

* Host OS version: Windows 7 Professional SP1 

* Host OS architecture: x86\_64

* Host CPU model: Intel Core i7-2600K

* Host RAM size: 16 GB

* (Optional) Host computer model:

**To Reproduce**

For example, please run following instruction.

mov ch, [MEM_MMIO]
8a 2e xx xx

Here, MEM_MMIO is the address of memory mapped IO.
In this case, the value will be set to BP (Low) in HAXM.

**Expected Behavior**

Expectation: 
The value should be set in CH.

**Reproducibility**

100% (Any case when using MMIO address)

**Additional context**

I don't know if this fix is appropriate, I got the expected results by modifying vcpu_read_gpr and vcpu_write_gpr in vcpu.c as follows:

```
static uint64_t vcpu_read_gpr(void *obj, uint32_t reg_index, uint32_t size)
{
	struct vcpu_t *vcpu = obj;
	uint64_t value = 0;
	if (reg_index >= 16) {
		hax_panic_vcpu(vcpu, "vcpu_read_gpr: Invalid register index\n");
		return 0;
	}
	if (size == 1) {
		memcpy(&value, (UINT8*)(&vcpu->state->_regs[reg_index & 0x3]) + (reg_index >> 2), size);
		return value;
	}
	memcpy(&value, &vcpu->state->_regs[reg_index], size);
	return value;
}
static void vcpu_write_gpr(void *obj, uint32_t reg_index, uint64_t value,
	uint32_t size)
{
	struct vcpu_t *vcpu = obj;
	if (reg_index >= 16) {
		hax_panic_vcpu(vcpu, "vcpu_write_gpr: Invalid register index\n");
		return;
	}
	if (size == 1) {
		memcpy((UINT8*)(&vcpu->state->_regs[reg_index & 0x3]) + (reg_index >> 2), &value, size);
		return;
	}
	memcpy(&vcpu->state->_regs[reg_index], &value, size);
}
```

**Comment by AlexAltea** [comment]
Thanks a lot for your report! This is indeed an issue, though the appropriate way to fix it would be inside the emulator, i.e. `emulate.c`, rather thanin the implementation of the Emulator<->VCPU interface at `vcpu.c` (or `test_emulator.cpp`).

I've fixed this issue on: https://github.com/kryptoslogic/haxm/commit/1a14b89805408b444e3733eb0eec4e0b1cd7c467. This patch depends on #185 to be merged first, so I'll submit this patch as pull request right afterwards, but feel free to test the [*fix-highbyte* branch](https://github.com/kryptoslogic/haxm/tree/fix-highbyte) already if you want.

**Comment by SimK98l** [comment]
Thank you for handling the issue promptly. This patch is working as expected.

**Comment by AlexAltea** [comment]
@SimK98l This should have been fixed in #185. Can you confirm whether the issue is solved?

**Comment by SimK98l** [comment]
I confirmed that the problem was solved after merging #185. Thank you!
